### PR TITLE
Handle None as defaults in configs and kwargs

### PR DESCRIPTION
**Pull request by arunkannawadi**
Hi Mike, the main intention of this PR is to be able to generate a `BinnedCorr2` object by passing all four of `nbins`, `min_sep`, `max_sep`, `bin_size` but with exactly one of them set `None`. I've made the changes for `BinnedCorr3` as well and modified some helper functions to not check if a `key` exists in a config/kwarg dict but instead check if it has been explicitly set. I'm assuming that nowhere in `treecorr` `None` has a special meaning.

### CODE REVIEW

**Comment by rmjarvis** [comment]
Thanks Arun. This seems sensible to me.  One thing I'd like to see is a new unit test that fails with the old version of the code, but passes now with your changes.

**Comment by rmjarvis** [comment]
Also, it looks like lots of the current unit tests are failing, so there is some backwards incompatibility in this for you to figure out.

**Comment by arunkannawadi** [comment]
Hoo, that's a lot of failures! I'll look into the errors and write aa new  unit test as well.

**Comment by arunkannawadi** [comment]
I didn't add a new test case per se, but modified the existing ones so that they'd fail without the code changes in the PR. Tests passed in my fork:
https://github.com/arunkannawadi/TreeCorr/runs/6435598018?check_suite_focus=true

**Comment by arunkannawadi** [comment]
I think there 1 or 2 lines that are not covered by the tests. I didn't run it locally, so I'll wait for codecov to pick it up when you run the tests here and update the tests to maintain the 100% coverage.

File: tests/test_gg.py [hunk 1]
```diff
@@ -431,7 +431,7 @@ def test_gg():
     g2 = -gamma0 * np.exp(-r2/2.) * (2.*x*y)/r0**2
 
     cat = treecorr.Catalog(x=x, y=y, g1=g1, g2=g2, x_units='arcmin', y_units='arcmin')
-    gg = treecorr.GGCorrelation(bin_size=0.1, min_sep=1., max_sep=100., sep_units='arcmin',
+    gg = treecorr.GGCorrelation(bin_size=0.1, min_sep=1., max_sep=100., nbins=None, sep_units='arcmin',
```
Comment (rmjarvis) [inline tests/test_gg.py:434]: Don't do this.  Leave the old tests alone, since they still need to work the same way they used to.  Please instead add a new test using None to make sure it is valid and does the same thing.

E.g. after what is currently here (in the old version), add:
```
# Using None rather than omiting nbins is equivalent.
gg2 = treecorr.GGCorrelation(bin_size=0.1, min_sep=1., max_sep=100., nbins=None, sep_units='arcmin')
gg2.process(cat)
assert gg2 == gg
```
I believe this would fail in 4.2, but should work with your update.  Likewise for the other tests you changed.

File: tests/test_config.py [hunk 2]
```diff
@@ -435,17 +446,37 @@ def test_merge():
...
     assert config2['ra_col'] == 'alpha2000'
 
+    # ... except when the value in kwargs is None.
+    kwargs = {'bin_size': 0.06, 'nbins': None}
+    config2 = treecorr.config.merge_config(config1, kwargs, treecorr.BinnedCorr2._valid_params)
+    assert config2['bin_size'] == 0.06
+    assert config2['nbins'] == config1['nbins']
+    assert kwargs['nbins'] == config2['nbins']
+    assert kwargs['nbins'] is not None
+    # Note: config2 is not in a valid state to be used to initialize BinnedCorr2.
```
Comment (arunkannawadi) [inline tests/test_config.py:461]: I'd like your opinions on this behavior, Mike. Essentially, one cannot unset a config by setting it as None explicitly in kwargs. If this is not a desired behavior, I'm happy to make changes to not make an exception for `None`.

File: tests/test_gg.py (hunk 1 above)
Comment (arunkannawadi) [reply tests/test_gg.py]: Looks like `gg2 == gg` doesn't hold true at machine precision even when they are initialized exactly in the same manner: https://colab.research.google.com/drive/1gXNzl0EOr8zui7luENpeQt74AOI-xPy5?usp=sharing

I'll expand the assertion into approximate comparison of all its attributes listed in `__eq__`.

Comment (rmjarvis) [reply tests/test_gg.py]: The problem is threading changing the order of operations randomly.  If you do both with num_threads=1, it should come out exact to machine precision.
```
gg2 = treecorr.GGCorrelation(bin_size=0.1, min_sep=1., max_sep=100., nbins=None, sep_units='arcmin')
gg2.process(cat, num_threads=1)  # Make sure OpenMP doesn't change order of operations.
gg.process(cat, num_threads=1)
assert gg2 == gg
```

File: tests/test_config.py (hunk 2 above)
Comment (rmjarvis) [reply tests/test_config.py:461]: Hm.  Good point.  This should do whatever is least surprising to the user.  And I think if a user put `config=config, bin_size=0.06, nbins=None` explicitly in the kwargs, they would expect it to override whatever `nbins` was in the config.  So probably we should switch that to work the other way.
```
    assert config2['nbins'] is None
```

File: tests/test_config.py [hunk 3]
```diff
@@ -435,17 +446,37 @@ def test_merge():
...
 
+    # ... even when when the value in kwargs is None.
+    kwargs = {'bin_size': 0.06, 'nbins': None}
+    config2 = treecorr.config.merge_config(config1, kwargs, treecorr.BinnedCorr2._valid_params)
+    assert config2['bin_size'] == 0.06
+    assert config2['min_sep'] == config1['min_sep']
+    assert config2['max_sep'] == config1['max_sep']
+    assert kwargs['nbins'] is None
+    assert config2['nbins'] is None
+
```
Comment (arunkannawadi) [inline tests/test_config.py:462]: Okay, I think it now does what a user would intuitively expect, and the tests seem to pass.

**Review by rmjarvis — APPROVED** [review]
Looks good.  Thanks Arun!
